ENH: Add ndmax param for numpy.array

pdnm opened the pull request:
This PR implements the proposal in #5933.

The parameter makes it possible to prevent `array` from digging deeper into the input, even if the lengths at the the next level are all equal.

Example:
```python
np.array([[1, 2], [3, 4]], ndmax=1)
# array([list([1, 2]), list([3, 4])], dtype=object)
```

eric-wieser commented on numpy/__init__.pxd:
```diff
@@ -585,7 +585,7 @@ cdef extern from "numpy/arrayobject.h":
     dtype PyArray_DescrNewByteorder (dtype, char)
     object PyArray_IterAllButAxis (object, int *)
-    #object PyArray_CheckFromAny (object, dtype, int, int, int, object)
+    #object PyArray_CheckFromAny (object, dtype, int, int, int, int, object)
```
Unfortunately we can't add arguments to existing API functions, that breaks users.

pdnm replied:
Fixed

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1929,7 +1929,149 @@ PyArray_FromAny(PyObject *op, PyArray_Descr *newtype, int min_depth,
+            Py_XDECREF(newtype);
+            ret = NULL;
+        }
+        else if (max_depth != 0 && PyArray_NDIM(arr) > max_depth) {
```
Zero is a valid depth - we probably need a different value to mean "not specified", perhaps -1.

pdnm replied:
@eric-wieser Do you mean the `max_depth` or the `ndmax` argument? I duplicated the function `PyArray_FromAny` to add a new parameter, so zero-`max_depth` was already there. Anyway I'm gonna refactor the code to avoid duplication

eric-wieser replied:
What I mean is, I'd like to see a test for `np.array([1], ndmax=0).ndim == 0`

pdnm replied:
@eric-wieser I made it so and added the test

Commit:
ENH: Add ndmax param for numpy.array

The ndmax param specifies the maximum depth at which the multiarray
is created from the source object.

eric-wieser commented on numpy/core/src/multiarray/multiarraymodule.c:
```diff
@@ -1666,13 +1666,14 @@ _array_fromobject(PyObject *NPY_UNUSED(ignored), PyObject *args, PyObject *kws)
 
 full_path:
-    if (!PyArg_ParseTupleAndKeywords(args, kws, "O|O&O&O&O&i:array", kwd,
+    if (!PyArg_ParseTupleAndKeywords(args, kws, "O|O&O&O&O&ii:array", kwd,
```
```suggestion
    if (!PyArg_ParseTupleAndKeywords(args, kws, "O|O&O&O&O&$ii:array", kwd,
```
Let's make these keyword-only

pdnm replied:
Done

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1912,10 +1912,21 @@ PyArray_GetArrayParamsFromObject(PyObject *op,
 {
     /*
-     * This is the main code to make a NumPy array from a Python
-     * Object.  It is called from many different places.
+     * This is a duplication of PyArray_FromAny to support the maxdims option. 
      */
```
I think you probably want to put back the old comment.

pdnm replied:
Fixed

eric-wieser commented on numpy/core/src/multiarray/ctors.c:
```diff
@@ -1912,10 +1912,21 @@ PyArray_GetArrayParamsFromObject(PyObject *op,
+ * Steals a reference to newtype --- which can be NULL
+ */
+NPY_NO_EXPORT PyObject *
+PyArray_FromAny_MaxDim(PyObject *op, PyArray_Descr *newtype, int maxdims,
+                int min_depth, int max_depth, int flags, PyObject *context)
```
What's the difference between `maxdims` and `max_depth`?

pdnm replied:
`max_depth` is like a guard, if the discovered depth is more than that, the creation fails

eric-wieser reviewed (commented):
Thanks for the patch. I'm not yet sure whether a `ndmax` argument to `array` is the right solution, but this patch does a nice job of showing it's a straighforward one.

At some point I'll read through the original issue and come back to this.

seberg commented:
Thanks for the PR. Sorry for being late to the party, but I realize there is not actually much overlap with my PR, so its fine.

We have to figure out what the exact API is that we want, e.g. I could imagine just fixing `ndim=3` instead of `ndmax` (but then you have to decide whether you allow broadcasting the input to the output).

One thing that probably currently fails is if the input is already an array(-like), i.e. we should probably pass in `max_depth` and `maxdims` from python.

seberg commented:
One API consideration. If we add this argument (whatever it is) we should deprecate ragged arrays for objects as well probably (although that can be a followup of course). That may be argument for a "maximum" as opposed to a fixed number of dims though.

pdnm commented:
> Thanks for the PR. Sorry for being late to the party, but I realize there is not actually much overlap with my PR, so its fine.
> 
> We have to figure out what the exact API is that we want, e.g. I could imagine just fixing `ndim=3` instead of `ndmax` (but then you have to decide whether you allow broadcasting the input to the output).
> 
> One thing that probably currently fails is if the input is already an array(-like), i.e. we should probably pass in `max_depth` and `maxdims` from python.

Indeed it currently fails if the input is an array-like.
I imagine `ndim=3` could be archived by the users with `np.array(obj, ndmin=3, ndmax=3)`

mattip commented:
> What "ragged arrays for objects" do you mean?

See [NEP 34](https://numpy.org/neps/nep-0034-infer-dtype-is-object.html)

mattip commented:
Very outdated, and mostly covered by NEP 34. We may want to continue this as an idea for consideration at some point, but would need to expose the discovery function.
